pkg/test/integration/common/framework.go

```diff
@@ -0,0 +1,772 @@
+			ginkgo.It("should not lead to errors and add futher 3 nodes to target cluster", func() {
+
+				retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					machineDployment, _ := c.ControlCluster.McmClient.MachineV1alpha1().MachineDeployments(controlClusterNamespace).Get("test-machine-deployment", metav1.GetOptions{})
+					machineDployment.Spec.Replicas = 6
+					_, updateErr := c.ControlCluster.McmClient.MachineV1alpha1().MachineDeployments(controlClusterNamespace).Update(machineDployment)
+					return updateErr
+				})
+				ginkgo.By("Checking for errors")
+				gomega.Expect(retryErr).NotTo(gomega.HaveOccurred())
+				// check whether all the expected nodes are ready
+				ginkgo.By("Checking number of ready nodes are 6 more than initial")
+				gomega.Eventually(c.TargetCluster.GetNumberOfNodes, 180, 5).Should(gomega.BeNumerically("==", initialNodes+6))
+				gomega.Eventually(c.TargetCluster.GetNumberOfReadyNodes, 180, 5).Should(gomega.BeNumerically("==", initialNodes+6))
+			})
+
+		})
+		ginkgo.Context("scale-down with replicas=2", func() {
+			// rapidly scaling back to 2, leading to freezing and unfreezing
+			// check for freezing and unfreezing of the machine due to rapidly scale up and scale down in the logs of mcm
+
+			ginkgo.It("should not lead to errors and remove 4 nodes from target cluster", func() {
+				retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					machineDployment, _ := c.ControlCluster.McmClient.MachineV1alpha1().MachineDeployments(controlClusterNamespace).Get("test-machine-deployment", metav1.GetOptions{})
+					machineDployment.Spec.Replicas = 2
+					_, updateErr := c.ControlCluster.McmClient.MachineV1alpha1().MachineDeployments(controlClusterNamespace).Update(machineDployment)
+					return updateErr
+				})
+				ginkgo.By("Checking for errors")
+				gomega.Expect(retryErr).NotTo(gomega.HaveOccurred())
+
+				ginkgo.By("Checking number of ready nodes are 2 more than initial")
+				gomega.Eventually(c.TargetCluster.GetNumberOfNodes, 300, 5).Should(gomega.BeNumerically("==", initialNodes+2))
+				gomega.Eventually(c.TargetCluster.GetNumberOfReadyNodes, 300, 5).Should(gomega.BeNumerically("==", initialNodes+2))
+			})
+			ginkgo.It("should freeze and unfreeze machineset temporarily", func() {
+				if mcsession == nil {
+					// controllers running in pod
+					// Create log file from container log
+					outputFile, err := helpers.RotateLogFile(mcmLogFile)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					ginkgo.By("Reading container log is leading to error")
+					mcmPod, err := c.ControlCluster.Clientset.CoreV1().Pods(controlClusterNamespace).List(metav1.ListOptions{})
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					readCloser, err := c.ControlCluster.Clientset.CoreV1().
+						Pods(controlClusterNamespace).
+						GetLogs(mcmPod.Items[0].Name, &coreV1.PodLogOptions{
+							Container: "machine-controller-manager",
+						}).Stream()
```

Comment:
Okay, in Line 600, all the pods in the control namespace are listed and in line 602, the logs from first pod in the list is tried to be streamed. How are you sure that the first pod (mcmPod.Items[0]) is an MCM pod? I think, we have to filter out the MCM pod here.

I see an error in while running as below.

```
Machine controllers test machine deployment resource scale-down with replicas=2
  should freeze and unfreeze machineset temporarily
  /Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/vendor/github.com/gardener/machine-controller-manager/pkg/test/integration/common/framework.go:593
STEP: Checking nodes in target cluster are healthy
STEP: Reading container log is leading to error
• Failure [0.872 seconds]
Machine controllers test
/Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/test/integration/controller/controller_test.go:45
  machine deployment resource
  /Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/vendor/github.com/gardener/machine-controller-manager/pkg/test/integration/common/framework.go:540
    scale-down with replicas=2
    /Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/vendor/github.com/gardener/machine-controller-manager/pkg/test/integration/common/framework.go:575
      should freeze and unfreeze machineset temporarily [It]
      /Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/vendor/github.com/gardener/machine-controller-manager/pkg/test/integration/common/framework.go:593
      Unexpected error:
          <*errors.StatusError | 0xc0004f8820>: {
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "container machine-controller-manager is not valid for pod cert-controller-manager-6bc5f9d784-phrdn",
                  Reason: "BadRequest",
                  Details: nil,
                  Code: 400,
              },
          }
          container machine-controller-manager is not valid for pod cert-controller-manager-6bc5f9d784-phrdn
      occurred
      /Users/i538135/go/src/github.com/gardener/machine-controller-manager-provider-aws/vendor/github.com/gardener/machine-controller-manager/pkg/test/integration/common/framework.go:607
``` 

I guess, my hunch is correct!

Reply:
Thanks for pointing this out. It has been fixed now.